File: services/agent-api/app/api/routes/strategy.py
"""Strategy learning API routes."""

import uuid

from fastapi import APIRouter, Query

from app.schemas.common import err, ok, paginated_ok
from app.services import strategy_service

router = APIRouter(tags=["Strategy"])


@router.get("/reranker-training-examples")
def list_reranker_training_examples(page: int = Query(1, ge=1), page_size: int = Query(20, ge=1, le=100), memory_id: str | None = None):
    result = strategy_service.list_reranker_examples(page, page_size, memory_id=uuid.UUID(memory_id) if memory_id else None)
    return paginated_ok(result["items"], page, page_size, result["total"])


@router.post("/reranker-training-examples")
def create_reranker_training_example(body: dict):
    try:
        return ok(strategy_service.create_reranker_example(body))
    except ValueError as exc:
        return err("RERANKER_SAMPLE_INVALID", str(exc))


@router.post("/reranker-training-examples/from-feedback/{feedback_event_id}")
def create_reranker_training_example_from_feedback(feedback_event_id: str, body: dict | None = None):
    row = strategy_service.build_reranker_example_from_feedback(uuid.UUID(feedback_event_id), body)
    return ok(row) if row else err("NOT_FOUND", "Feedback event not found")


@router.get("/memory-reranker-runs")
def list_memory_reranker_runs(page: int = Query(1, ge=1), page_size: int = Query(20, ge=1, le=100), trace_run_id: str | None = None):
    result = strategy_service.list_memory_reranker_runs(page, page_size, trace_run_id=uuid.UUID(trace_run_id) if trace_run_id else None)
    return paginated_ok(result["items"], page, page_size, result["total"])


@router.post("/memory-reranker-runs")
def create_memory_reranker_run(body: dict):
    try:
        return ok(strategy_service.create_memory_reranker_run(body))
    except ValueError as exc:
        return err("INVALID_LEARNING_MODE", str(exc))


@router.post("/memory-reranker-models/train")
def train_memory_reranker(body: dict):
    try:
        companion_id = uuid.UUID(str(body["companion_id"]))
        return ok(strategy_service.train_memory_reranker(companion_id))
    except (KeyError, ValueError) as exc:
        return err("RERANKER_TRAINING_INVALID", str(exc))


@router.get("/presence-policy-feedback-samples")
def list_presence_policy_feedback_samples(page: int = Query(1, ge=1), page_size: int = Query(20, ge=1, le=100), presence_opportunity_id: str | None = None):
    result = strategy_service.list_presence_feedback_samples(page, page_size, presence_opportunity_id=uuid.UUID(presence_opportunity_id) if presence_opportunity_id else None)
    return paginated_ok(result["items"], page, page_size, result["total"])


@router.post("/presence-policy-feedback-samples")
def create_presence_policy_feedback_sample(body: dict):
    return ok(strategy_service.create_presence_feedback_sample(body))


@router.get("/presence-policy-runs")
def list_presence_policy_runs(page: int = Query(1, ge=1), page_size: int = Query(20, ge=1, le=100), trace_run_id: str | None = None):
    result = strategy_service.list_presence_policy_runs(page, page_size, trace_run_id=uuid.UUID(trace_run_id) if trace_run_id else None)
    return paginated_ok(result["items"], page, page_size, result["total"])


@router.post("/presence-policy-runs")
def create_presence_policy_run(body: dict):
    try:
        return ok(strategy_service.create_presence_policy_run(body))
    except ValueError as exc:
        return err("INVALID_LEARNING_MODE", str(exc))

File: apps/web/lib/api/strategy.ts
import { api, queryString, type QueryParams } from "./client";
import type { LearningMode, PaginatedItems } from "@/lib/types";

export interface RerankerTrainingExample {
  id: string;
  memory_id?: string | null;
  feedback_event_id?: string | null;
  label: number;
  source_type: string;
  feature_json?: Record<string, unknown>;
}

export interface MemoryRerankerRun {
  id: string;
  trace_run_id?: string | null;
  learning_mode: LearningMode;
  status: string;
  candidate_memory_ids: string[];
  selected_memory_ids: string[];
  score_json?: Record<string, unknown>;
  explanation_json?: Record<string, unknown>;
}

export interface PresencePolicyFeedbackSample {
  id: string;
  presence_opportunity_id?: string | null;
  feedback_event_id?: string | null;
  action_taken: string;
  reward: number;
  feature_json?: Record<string, unknown>;
}

export interface PresencePolicyRun {
  id: string;
  trace_run_id?: string | null;
  presence_opportunity_id?: string | null;
  learning_mode: LearningMode;
  action_space: string[];
  selected_action: string;
  reward_prediction?: number | null;
  explanation_json?: Record<string, unknown>;
}

export function listRerankerTrainingExamples(params?: QueryParams) {
  return api.get<PaginatedItems<RerankerTrainingExample>>(`/reranker-training-examples${queryString(params)}`);
}
export function createRerankerTrainingExample(data: Partial<RerankerTrainingExample>) {
  return api.post<RerankerTrainingExample>("/reranker-training-examples", data);
}
export function createRerankerTrainingExampleFromFeedback(feedbackEventId: string, data?: Record<string, unknown>) {
  return api.post<RerankerTrainingExample>(`/reranker-training-examples/from-feedback/${feedbackEventId}`, data);
}
export function listMemoryRerankerRuns(params?: QueryParams) {
  return api.get<PaginatedItems<MemoryRerankerRun>>(`/memory-reranker-runs${queryString(params)}`);
}
export function createMemoryRerankerRun(data: Partial<MemoryRerankerRun> & { allow_active?: boolean }) {
  return api.post<MemoryRerankerRun>("/memory-reranker-runs", data);
}
export function listPresencePolicyFeedbackSamples(params?: QueryParams) {
  return api.get<PaginatedItems<PresencePolicyFeedbackSample>>(`/presence-policy-feedback-samples${queryString(params)}`);
}
export function createPresencePolicyFeedbackSample(data: Partial<PresencePolicyFeedbackSample>) {
  return api.post<PresencePolicyFeedbackSample>("/presence-policy-feedback-samples", data);
}
export function listPresencePolicyRuns(params?: QueryParams) {
  return api.get<PaginatedItems<PresencePolicyRun>>(`/presence-policy-runs${queryString(params)}`);
}
export function createPresencePolicyRun(data: Partial<PresencePolicyRun> & { allow_active?: boolean }) {
  return api.post<PresencePolicyRun>("/presence-policy-runs", data);
}
